pkg/edgedictionary/update.go

```diff
@@ -0,0 +1,56 @@
+package edgedictionary
+
+import (
+	"io"
+
+	"github.com/fastly/cli/pkg/common"
+	"github.com/fastly/cli/pkg/compute/manifest"
+	"github.com/fastly/cli/pkg/config"
+	"github.com/fastly/cli/pkg/errors"
+	"github.com/fastly/cli/pkg/text"
+	"github.com/fastly/go-fastly/fastly"
+)
+
+// UpdateCommand calls the Fastly API to describe a service.
+type UpdateCommand struct {
+	common.Base
+	manifest manifest.Data
+	Input    fastly.UpdateDictionaryInput
+}
+
+// NewUpdateCommand returns a usable command registered under the parent.
+func NewUpdateCommand(parent common.Registerer, globals *config.Data) *UpdateCommand {
+	var c UpdateCommand
+	c.Globals = globals
+	c.manifest.File.Read(manifest.Filename)
+	c.CmdClause = parent.Command("update", "Update name of dictionary on a Fastly service version").Alias("get")
+	c.CmdClause.Flag("service-id", "Service ID").Short('s').StringVar(&c.manifest.Flag.ServiceID)
+	c.CmdClause.Flag("version", "Number of service version").Required().IntVar(&c.Input.Version)
+	c.CmdClause.Flag("name", "Old name of Dictionary").Short('n').Required().StringVar(&c.Input.Name)
+	c.CmdClause.Flag("new-name", "New name of Dictionary").Required().StringVar(&c.Input.NewName)
+	return &c
+}
+
+// Exec invokes the application logic for the command.
+func (c *UpdateCommand) Exec(in io.Reader, out io.Writer) error {
+	serviceID, source := c.manifest.ServiceID()
+	if source == manifest.SourceUndefined {
+		return errors.ErrNoServiceID
+	}
+	c.Input.Service = serviceID
+
+	d, err := c.Globals.Client.UpdateDictionary(&c.Input)
+	if err != nil {
+		return err
+	}
+
+	text.Success(out, "Updated dictionary %s to %s (service %s version %d)", c.Input.Name, d.Name, d.ServiceID, d.Version)
+
+	if c.Globals.Verbose() {
+		text.Output(out, "Service ID: %s", d.ServiceID)
+		text.Output(out, "Version: %d", d.Version)
+		text.PrintDictionary(out, "", d)
+	}
+
+	return nil
+}
```

Comment:
This logic will need to change now that we've upgraded to `go-fastly/v2` in #169. Specifically, all update structs now use pointers in their field values.  If you look at a foo/update.go file in that PR you will see the new2 pattern which uses our optional types.

Reply:
I'm noticing these don't use kingpin's `Required()` and instead relies on its own `WasSet` to fail if the value isn't set. Is that for a future migration off of kingpin? Otherwise, it seems like unnecessary code.

Reply:
@benzvan 

There was a change in Go-Fastly 2.0.0 where [basic type](https://tour.golang.org/basics/11) fields that are _optional_ on 'update' input structs, were changed to pointers using some new [helper functions](./fastly/basictypes_helper.go). 

This was done to avoid unexpected behaviours when dealing with their zero value (see [this reference](https://willnorris.com/2014/05/go-rest-apis-and-pointers/) for more details).

So here's where things get a bit complicated (thanks Kingpin), and I might get some of the details wrong as I'm not overly familiar with Kingpin at this point myself. So @phamann feel free to chime in and correct me at any point.

Because of that change in Go-Fastly, we can no longer have Kingpin assign a flag value directly to the relevant Go-Fastly struct field because it's unable to write to a pointer. This  means these Go-Fastly optional pointer fields now need to be populated indirectly by CLI logic.

To do that we define the optional fields on command struct using the `common.OptionalString` type (or [one of the other 'Optional<T>' types](https://github.com/fastly/cli/blob/782b9ed756761555fe6c45c6cc9c2c1480ca64fc/pkg/common/command.go#L71-L105) we have).

We now assign the CLI flag value to the `Value` field on the `common.OptionalString` type and we make Kingpin execute a 'callback' function using its `Action()` method to tell us the flag was provided (as I don't believe Kingpin natively gives us a way to tell if a flag was specified by a user).

Example:

```go
type UpdateCommand struct {
  ...

  Comment common.OptionalString
}
```

```diff
- c.CmdClause.Flag("comment", "A descriptive note").StringVar(&c.updateInput.Comment)
+ c.CmdClause.Flag("comment", "A descriptive note").Action(c.Comment.Set).StringVar(&c.Comment.Value)
```

You'll see the callback we execute (`c.Comment.Set`) in this case references a `common.Optional` struct method called `Set` that is embedded into the `common.OptionalString` type.

What that does is store off whether the flag value was actually set using a `WasSet` field. 

Later on in the relevant receiver's `Exec` method we can now set the Go-Fastly struct field using our type-to-pointer helper functions.

Example:

```diff
- if c.updateInput.Comment == "" { c.updateInput.Comment = d.Comment }
+ if c.Comment.WasSet { c.updateInput.Comment = fastly.String(c.Comment.Value) }
```

Reply:
Thank you for that explanation of these changes. I'm still unclear about the `Required` function. In my updated code, I have this:

```
c.CmdClause.Flag("new-name", "New name of Dictionary").Required().Action(c.newname.Set).StringVar(&c.newname.Value)
```
So I'm still using `Flag.Required()` to make sure the `--new-name` flag was passed but then also using the `Action` function to call `c.newname.Set`.

In my tests, I verify that it fails if `--new-name` is not passed but I don't use `c.newname.WasSet` because if it gets past the command line parsing, it should be set.

Maybe this is partially a stylistic change and I should drop the `Required()` from that line and add the `WasSet` below?